xbout/grid.py

```diff
@@ -0,0 +1,74 @@
+from pathlib import Path
+from warnings import warn
+
+import xarray as xr
+
+from . import geometries
+from .utils import _set_attrs_on_all_vars, _check_filetype, _separate_metadata
+
+
+def open_grid(gridfilepath='./grid.nc', geometry=None, ds=None, quiet=False):
+    """
+    Opens a BOUT++ grid file.
+
+    Opens a BOUT++ grid file (as generated by Hypnotoad etc.). Then sets
+    coordinates depending on what's in the file, and moves grid data to the
+    attributes dict.
+
+    Parameters
+    ----------
+    gridfilepath : str, optional
+    geometry : str, optional
+        The geometry type of the grid data. This will specify what type of
+        coordinates to add to the dataset, e.g. 'toroidal' or 'cylindrical'.
+
+        If not specified then will attempt to read it from the file attrs.
+        If still not found then a warning will be thrown, which can be
+        suppressed by passing `quiet`=True.
+
+        To define a new type of geometry you need to use the
+        `register_geometry` decorator.
+    ds : xarray.Dataset, optional
+        BOUT dataset to merge grid information with.
+        Leave unspecified if you just want to open the grid file alone.
+
+    Returns
+    -------
+    ds : xarray.Dataset
+
+    """
+
+    gridfilepath = Path(gridfilepath)
+    grid = xr.open_dataset(gridfilepath, engine=_check_filetype(gridfilepath))
+
+    # TODO find out what 'yup_xsplit' etc are in the doublenull storm file John gave me
+    # For now drop any variables with extra dimensions
+    variables = list(grid.variables)
+    vars_to_drop = [var for var in variables
+                    if not all(dim in ['t', 'x', 'y', 'z']
+                               for dim in grid[var].dims)]
```

Comment:
This is clever, but unfortunately difficult to read! Please could you rewrite this to be a bit simpler?